### PR TITLE
DatasetBuilder and DatasetBenchmarker

### PR DESCRIPTION
**Pull request by ngreenwald**
This PR creates a DatasetBuilder and DatasetBenchmarker class for constructing and evaluating model accuracy on a range of different datasets. 

**DatasetBuilder**
1. The DatasetBuilder assumes that the data is stored in a folder structure. Each folder corresponds to a single experiment, with a single metadata file. However, there can be multiple NPZ files within each experiment representing different caliban jobs, different days of the same experiment, etc. This will eventually be replaced with a call to deepcell-datasets.
2. The motivation for this class was the lack of metadata that is retained in our current NPZ setup. In particular, we can't map individual images within the array back to their phenotype information. This class extends the dict{} structure we've been using before with two additional keys. tissue_list is a list of the same length as the number of images with the tissue type of each image. platform_list is the same but for image platform.
3.  The trigger for creating a dataset is calling the build_dataset function. However, this method doesn't do much work itself, and is mostly in charge of calling the accessory functions. 
4. `_load_experiment` loads a single folder of data. `_load_all_experiments` loads the entire dataset, and performs the appropriate train/va/test split of the data as it is being loaded in. `_subset_data_dict` takes a dictionary containing X, y, tissue, and platform type, and subsets it to include the specified tissue and platform types. `_reshape_dict` reshapes each of the dicts to account for cropping and resizing of the training data. `_clean_labels` cleans up the labels prior to creating the final dict. 

**DatasetBenchmarker**
This class takes a set of true labels, predicted labels, and a list of the tissue and platform type for each image, and computes metrics across each unique tissue and platform type within the dataset. I was able to abstract this one a bit more, and it's fairly clear how additional categories would be able to easily slot in (i.e. culture condition, resolution, etc). 

**Other changes**
Adds a new function, create_train_val_test_split, which creates the specified split in the data. 
Removes combine_npz_files, which was a filler for the DataLoader

**Areas in need of TLC**
DatasetBuilder._load_all_experiments has 12 different lists [X, y, tissue_list, platform_list] x [train, val, test]. Each of these have a separate _batch version which is used as a working version during the for loop, which then gets appended to the main list, and finally concatenated. The logic is 100% identical between them. What would be a better way to handle this? I was thinking maybe zipping them all together?

DatasetBuilder._reshape_dict has almost the exact same code for reshape_by_img and reshape_by_tissue. The difference is that the first loops over each image, whereas the second loops over each unique tissue type. But the machinery is identical within them. 

build.train_val_test_split currently fails if there aren't enough images to be divided among the three classes. What is the best way to handle this?

### CODE REVIEW

**Comment by ngreenwald** [comment]
Updated the data_splitting function to handle small datasizes that don't fit into assigned splits

File: caliban_toolbox/utils/misc_utils.py [hunk 1]
```diff
@@ -119,3 +119,19 @@ def list_npzs_folder(npz_dir):
...
+        base_dir: directory with folders
+
+    Returns:
+        list of folders in base_dir, empty if None
+    """
+
+    files = os.listdir(base_dir)
+    folders = [file for file in files if os.path.isdir(os.path.join(base_dir, file))]
+
+    return folders
```
Comment (willgraf) [inline caliban_toolbox/utils/misc_utils.py:137]: [This might be improved by using `os.walk` instead of `os.listdir`](https://stackoverflow.com/a/973488)

File: caliban_toolbox/DatasetBenchmarker.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            stats_dict: dictionary of benchmarking results
+        """
+
+        unique_ids = np.unique(category_ids)
+
+        # create dict to hold stats across each category
+        stats_dict = {}
+        for uid in unique_ids:
+            stats_dict[uid] = {}
+            category_idx = np.isin(category_ids, uid)
```
Comment (willgraf) [inline caliban_toolbox/DatasetBenchmarker.py:89]: This seems like its a boolean, not an index? How is this used?

**Comment by ngreenwald** [comment]
- Modified train_val_test_split so that all decisions are random, based on seed
- Raises a warning if data split has to be modified
- Adds a validate_dataset function to make sure each folder has an NPZ and metadata file
- Adds a get_metadata function to further abstract where metadata is stored
- combines repeated code into single _reshape_dict loop for resizing

File: caliban_toolbox/build.py [hunk 3]
```diff
@@ -43,14 +45,15 @@ def compute_cell_size(npz_file, method='median', by_image=True):
             the cell size across the entire npz is returned
 
     Returns:
-        average_sizes: list of typical cell size in NPZ
+        average_sizes: list of typical cell size in NPZ. If no cells, returns None
```
Comment (willgraf) [inline caliban_toolbox/build.py:48]:
```suggestion
        list: list of typical cell size in NPZ. If no cells, returns None.
```

The text before the colon is supposed to be the data type, not the var name.

File: caliban_toolbox/build.py [hunk 4]
```diff
@@ -43,14 +45,15 @@ def compute_cell_size(npz_file, method='median', by_image=True):
...
 
     Raises: ValueError if invalid method supplied
     Raises: ValueError if data does have len(shape) of 4
     """
 
-    valid_methods = set(['median', 'mean'])
-    if method.lower() not in valid_methods:
+    valid_methods = {'median', 'mean'}
+    method = method.lower()
```
Comment (willgraf) [inline caliban_toolbox/build.py:55]: You can cast as a string to make sure that `.lower()` doesn't throw an attribute error.

```suggestion
    method = str(method).lower()
```

File: caliban_toolbox/dataset_builder.py [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+                if median_cell_size is not None:
+                    resize_ratio = np.sqrt(resize_target / median_cell_size)
+                else:
+                    resize_ratio = 1
+
+                # resize the data
+                X_batch_resized, y_batch_resized = \
+                    reshape_training_data(X_data=X_batch, y_data=y_batch,
+                                          resize_ratio=resize_ratio, final_size=output_shape,
+                                          tolerance=resize_tolerance)
```
Comment (willgraf) [inline caliban_toolbox/dataset_builder.py:356]: No need for the `\` if the line breaks are inside a paranethesis.

```suggestion
                X_batch_resized, y_batch_resized = reshape_training_data(
                    X_data=X_batch, y_data=y_batch,
                    resize_ratio=resize_ratio, final_size=output_shape,
                    tolerance=resize_tolerance)
```

File: caliban_toolbox/dataset_builder.py [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+        platform_list = np.array(data_dict['platform_list'])
+        keep_idx = np.repeat(True, y.shape[0])
+        cleaned_y = np.zeros_like(y)
+
+        # TODO: remove one data QC happens in main toolbox pipeline
+        for i in range(y.shape[0]):
+            if relabel_hard:
+                y_new = label(y[i, :, :, 0])
+            else:
+                y_new, _, _ = relabel_sequential(y[i, :, :, 0])
```
Comment (willgraf) [inline caliban_toolbox/dataset_builder.py:405]: I haven't done the legwork, but I am still unconvinced this is important.

File: caliban_toolbox/dataset_builder.py [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+                if tissue not in self.all_tissues:
+                    raise ValueError('{} is not one of {}'.format(tissue, self.all_tissues))
+        elif tissues == 'all':
+            tissues = self.all_tissues
+        elif tissues in self.all_tissues:
+            tissues = [tissues]
+        else:
+            raise ValueError(
+                'tissues should be "all", one of {}, or a list '
+                'of acceptable tissue types'.format(self.all_tissues))
```
Comment (willgraf) [inline caliban_toolbox/dataset_builder.py:468]: These 2 bits of logic look so similar, it could probably be a `validate` function that works for both tissues and platforms.

**Comment by ngreenwald** [comment]
Changes:

- Added explicit checks for 4D data where it's currently required, abstracted away where it's not
- Added type checking for labeled data (must be int) for skimage functions to properly work
- Removed relabel_sequential in favor of label, which accomplishes same thing

File: caliban_toolbox/build_test.py [hunk 8]
```diff
@@ -167,100 +224,77 @@ def test_pad_image_stack():
...
-
-    combined_x, combined_y = build.combine_npz_files(npz_list=npz_resize_list,
-                                                     resize_ratios=resize_ratios,
-                                                     final_size=final_size)
-
-    # check that correct number of NPZs present
-    assert combined_x.shape[0] == np.sum(2 + 1 + 4)
+def test_train_val_test_split():
+    X_data = np.zeros((100, 5, 5, 3))
+    y_data = np.zeros((100, 5, 5, 1))
```
Comment (willgraf) [inline caliban_toolbox/build_test.py:229]: You may want to add another test that confirms reproducibility with a given seed.

File: caliban_toolbox/dataset_builder_test.py [hunk 9]
```diff
@@ -0,0 +1,615 @@
...
+# limitations under the License.
+# ==============================================================================
+import os
+import json
+import pytest
+
+import numpy as np
+from pathlib import Path
+
+
```
Comment (willgraf) [inline caliban_toolbox/dataset_builder_test.py:33]:
```suggestion
```

**Comment by ngreenwald** [comment]
- Added the option to specify different output sizes for each data split
- Added the option to resize by a constant value for all images, for example to for different resolutions
- Added check for split reproducibility with seed

**Comment by ngreenwald** [comment]
Added support for any combination of a list or tuple for various output_shape parameters

File: caliban_toolbox/dataset_builder.py [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+            platforms: which platforms to include. Must be either a list of platform types,
+                a single platform type, or 'all'
+            output_shape: output shape for dataset. Either a single tuple, in which case
+                train/va/test will all have same size, or a list of three tuples
+            resize: flag to control resizing the input data.
+                Valid arguments:
+                    - False. No resizing
+                    - float/int: Resizes all images by supplied value
+                    - by_tissue. Resizes by median cell size within each tissue type
+                    - by_image. REsizes by median cell size within each image
```
Comment (willgraf) [inline caliban_toolbox/dataset_builder.py:524]:
```suggestion
                    - by_image. Resizes by median cell size within each image
```

File: caliban_toolbox/dataset_builder.py [hunk 11]
```diff
@@ -0,0 +1,627 @@
...
+            # then we duplicate the labels in place to match expanded array size
+            tissue_list_new = [item for item in tissue_list for _ in range(multiplier)]
+            platform_list_new = [item for item in platform_list for _ in range(multiplier)]
+        else:
+            X_new, y_new, tissue_list_new, platform_list_new = [], [], [], []
+
+            if resize == 'by_tissue':
+                batch_ids = np.unique(tissue_list)
+            else:
+                batch_ids = np.arange(0, X.shape[0])
```
Comment (willgraf) [inline caliban_toolbox/dataset_builder.py:342]: Even though the input gets validated in a different file, I would feel better if we specifically checked the string value instead of inferring it in the else.

```suggestion
            if resize == 'by_tissue':
                batch_ids = np.unique(tissue_list)
            elif resize == 'by_image':
                batch_ids = np.arange(0, X.shape[0])
            else:
                raise ValueError('Invalid `resize` value: {}'.format(resize))
```

**Review by willgraf — CHANGES_REQUESTED** [review]

**Comment by ngreenwald** [comment]
For sure, always better to be safe

**Review by willgraf — APPROVED** [review]

**Review by MekWarrior — APPROVED** [review]

File: caliban_toolbox/dataset_builder.py [hunk 12]
```diff
@@ -0,0 +1,629 @@
...
+        Returns:
+            cleaned_dict: dictionary with cleaned labels
+        """
+        X, y = data_dict['X'], data_dict['y']
+        tissue_list = np.array(data_dict['tissue_list'])
+        platform_list = np.array(data_dict['platform_list'])
+        keep_idx = np.repeat(True, y.shape[0])
+        cleaned_y = np.zeros_like(y)
+
+        # TODO: remove one data QC happens in main toolbox pipeline
```
Comment (ngreenwald) [inline caliban_toolbox/dataset_builder.py:417]:
```suggestion
        # TODO: remove once data QC happens in main toolbox pipeline
```
